cmd/bosun/expr/funcs.go

```diff
@@ -787,3 +813,225 @@ func Transpose(e *State, T miniprofiler.Timer, d *Results, gp string) (*Results,
+		}
+		series := make(Series)
+		for _, v := range ts.Buckets {
+			series[time.Unix(v.Key/1000, 0).UTC()] = float64(v.DocCount) / ds.Seconds()
+		}
+		r.Results = append(r.Results, &Result{
+			Value: series,
+			Group: make(opentsdb.TagSet),
+		})
+		return r, nil
+	}
+	aggregation := elastic.NewTermsAggregation().Field(keys[len(keys)-1].Key)
+	aggregation = aggregation.SubAggregation("ts", ts)
+	if len(keys) > 1 {
+		for i := len(keys) - 2; i > -1; i-- {
+			aggregation = elastic.NewTermsAggregation().Field(keys[i].Key).SubAggregation("g_"+keys[i+1].Key, aggregation)
+		}
+	}
+	s = s.Aggregation("g_"+keys[0].Key, aggregation)
+	data, err := json.MarshalIndent(s.Source(), "", "\t")
+	fmt.Println(string(data))
+	if err != nil {
+		panic("error generating raw elastic query")
+	}
+	fmt.Println(string(data))
+	result, err := service.SearchSource(s).Do()
+	if err != nil {
+		return nil, err
+	}
+	top, ok := result.Aggregations.Terms("g_" + keys[0].Key)
+	if !ok {
+		return nil, fmt.Errorf("top key g_%v not found in result", keys[0].Key)
+	}
+	for _, b := range top.Buckets {
+		tags := make(opentsdb.TagSet)
+		key := fmt.Sprintf("%v", b.Key)
+		if keys[0].Pattern != nil && !keys[0].Pattern.MatchString(key) {
+			continue
+		}
+		tags[keys[0].Key] = key
+		desc(b, tags, 1, keys, r, e, ds)
+	}
+	return r, nil
+}
+
+func desc(b *elastic.AggregationBucketKeyItem, tags opentsdb.TagSet, depth int, keys []keyMatch, r *Results, e *State, ds opentsdb.Duration) {
+	if ts, found := b.DateHistogram("ts"); found {
+		if e.squelched(tags) {
+			return
+		}
+		series := make(Series)
+		for _, v := range ts.Buckets {
+			series[time.Unix(v.Key/1000, 0).UTC()] = float64(v.DocCount) / ds.Seconds()
+		}
+		r.Results = append(r.Results, &Result{
+			Value: series,
+			Group: tags.Copy(),
+		})
+		return
+	}
+	n, ok := b.Aggregations.Terms("g_" + keys[depth].Key)
+	if !ok {
+		panic(fmt.Sprintf("%v not found at depth %v", "g_"+keys[depth].Key, strconv.Itoa(depth)))
+	}
+	for _, item := range n.Buckets {
+		key := fmt.Sprintf("%v", item.Key)
+		if keys[depth].Pattern != nil && !keys[depth].Pattern.MatchString(key) {
+			continue
+		}
+		tags[keys[depth].Key] = key
+		desc(item, tags, depth+1, keys, r, e, ds)
+	}
+}
+
+func GenLSIndexes(host, index_root string, start, end time.Time) (string, error) {
+	resp, err := http.Get("http://" + host + "/_cat/indices")
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	lines := strings.Split(string(body), "\n")
+	existing_indicies := make(map[string]bool)
+	for _, line := range lines {
+		f := strings.Fields(line)
+		if len(f) > 1 {
+			existing_indicies[f[2]] = true
+		}
+	}
+	var indicies []string
+	for d := start; d.Before(end); d = d.AddDate(0, 0, 1) {
+		i := fmt.Sprintf("%s-%s", index_root, d.Format("2006.01.02"))
+		if _, ok := existing_indicies[i]; ok {
```

Comment:
My spidey sense suggests this loop should be merged with the loop on line 1020. That is, look through all the indicies only one time, and do the checking right then. Split on the `-`, check the index_root against [0], parse the time, and see if it's within range. If so, add it to indices.